lib/Parse/ParseExpr.cpp

```diff
@@ -3023,6 +3022,142 @@ ExprResult Parser::ParseBoundsExpression() {
   return Result;
 }
 
+// Recurse through types and substitute any TypedefType with TypeVariableType
+// as the underlying type.
+QualType Parser::SubstituteTypeVariable(QualType QT,
+  SmallVector<DeclRefExpr::GenericInstInfo::TypeArgument, 16> &typeNames) {
+  const Type *T = QT.getTypePtr();
+  switch (T->getTypeClass()) {
+  case Type::Pointer: {
+    const PointerType *pType = dyn_cast<PointerType>(T);
+    if (!pType) {
+      llvm_unreachable("Type cannot be empty when substituting type variable.");
```

Comment:
I think a better unreachable error message would be "dynamic cast failed unexpectedly."